packages/type-utils/src/builtinSymbolLikes.ts

```diff
@@ -105,7 +117,7 @@ export function isBuiltinTypeAliasLike(
 export function isBuiltinSymbolLike(
   program: ts.Program,
   type: ts.Type,
-  symbolName: string,
+  predicate: (symbolName: string) => boolean,
```

Comment:
[Refactor] Every one of the uses for this function does >=1 `===` check. How about using a `Set` instead to make it more clear what's happening?

```suggestion
  symbolNames: Set<string>
```

Reply:
By the way, these functions are already included in several public releases. Shouldn't we keep the already published signatures to avoid breaking changes? Perhaps we can add this new functionality as overload/union? Although these packages are not documented yet, not sure, just asking :smile:

Reply:
Predicates are more composable than sets and offer easier ergonomics; for sets; `(v) => symbolNameSet.has(v)` is just as readable as `symbolNameSet`, and for looking up a single value `(symbolName) => symbolName === 'asdf'` is in my opinion much easier to interpret than `new Set(['foo'])` which offers no context for what the set is for, so you end up needing either a comment or breaking it out into a constant. Finally, predicates can be more consistently applied; e.g., we already have predicates for `isBuiltinSymbolLikeRecurser`, so switching between predicates and sets for arguments requires constantly deciding which ones "tend to be used like sets". Furthermore, predicates portray the intent of the argument more clearly: a set has many functions besides just contains-checking, including iteration. Providing a predicate makes it clear the only part the library function needs is the yes/no answer on each element.

For @auvred 's comment; happy to add the overload/union; adds a little clutter though, so will wait for more discussion since I'm not familiar enough with the project to have any opinion for backwards compatibility vs tidiness for those helpers

Reply:
> backwards compatibility

Yup, this'll need to retain backwards compatibility. I haven't heard of anybody telling us they use it, but [a Sourcegraph query for TypeScript files referring to it](https://sourcegraph.com/search?q=context:global+%40typescript-eslint/type-utils+lang:TypeScript++-repo:%5Egithub%5C.com/typescript-eslint%28.*%29+-file:%28.*%29node_modules%28.*%29&patternType=standard&sm=1) shows a few public repos that do.

> `(v) => symbolNameSet.has(v)` is just as readable as `symbolNameSet`

I think this is subjective 😄 I find the latter more intuitive. In my mind I'm kind of implicitly applying the context from _"is built-in symbol like"_ to the last parameter. But that's just me.

> offers no context for what the set is for, so you end up needing either a comment or breaking it out into a constant

This is equally applicable to both a predicate and a set. For both, a descriptive parameter name like `allowedSymbolNames` or `isAllowedSymbolName` fixes the context issue.

> predicates portray the intent of the argument more clearly
> a set has many functions besides ...

I suppose `Pick<Set<String>, 'has'>` would be the most precise thing to take in. All we really want in any of this is to know whether a name is what we're looking for.

---


---

Reply:
> backwards compatibility

Yup, this'll need to retain backwards compatibility. I haven't heard of anybody telling us they use it, but [a Sourcegraph query for TypeScript files referring to it](https://sourcegraph.com/search?q=context:global+%40typescript-eslint/type-utils+lang:TypeScript++-repo:%5Egithub%5C.com/typescript-eslint%28.*%29+-file:%28.*%29node_modules%28.*%29&patternType=standard&sm=1) shows a few public repos that do.

> `(v) => symbolNameSet.has(v)` is just as readable as `symbolNameSet`

I think this is subjective 😄 I find the latter more intuitive. In my mind I'm kind of implicitly applying the context from _"is built-in symbol like"_ to the last parameter. But that's just me.

> offers no context for what the set is for, so you end up needing either a comment or breaking it out into a constant

This is equally applicable to both a predicate and a set. For both, a descriptive parameter name like `allowedSymbolNames` or `isAllowedSymbolName` fixes the context issue.

> predicates portray the intent of the argument more clearly
> a set has many functions besides ...

I suppose `Pick<Set<String>, 'has'>` would be the most precise thing to take in. All we really want in any of this is to know whether a name is what we're looking for.

---

Anyway, the issue of whether we use a set or predicate is not particularly important to me. It's a small implementation detail and if you feel strongly about it I don't mind just shrugging and seeing what happens.

Reply:
> backwards compatibility

Yup, this'll need to retain backwards compatibility. I haven't heard of anybody telling us they use it, but [a Sourcegraph query for TypeScript files referring to it](https://sourcegraph.com/search?q=context:global+%40typescript-eslint/type-utils+lang:TypeScript++-repo:%5Egithub%5C.com/typescript-eslint%28.*%29+-file:%28.*%29node_modules%28.*%29&patternType=standard&sm=1) shows a few public repos that do.

> `(v) => symbolNameSet.has(v)` is just as readable as `symbolNameSet`

I think this is subjective 😄 I find the latter more intuitive. In my mind I'm kind of implicitly applying the context from _"is built-in symbol like"_ to the last parameter. But that's just me.

> offers no context for what the set is for, so you end up needing either a comment or breaking it out into a constant

This is equally applicable to both a predicate and a set. For both, a descriptive parameter name like `allowedSymbolNames` or `isAllowedSymbolName` fixes the context issue.

> predicates portray the intent of the argument more clearly
> a set has many functions besides ...

I suppose `Pick<Set<String>, 'has'>` would be the most precise thing to take in. All we really want in any of this is to know whether a name is what we're looking for.

One flaw of functions is it leaks the implementation detail of _"this gets called up to once per symbol name"_, so folks can do weird things like:

```ts
isBuiltinSymbolLike(program, type, (name) => {
  someRegistry.add(name);
  if (someAttribute(name)) {
    context.report({ data: { name }, messageId: "someMessage", node: someNode });
  }
});
```

Just pointing this out as a counterpoint to the idea that a predicate adds more practical or semantic safety. I don't think we have to worry about anybody doing such an odd thing.

---

Anyway, the issue of whether we use a set or predicate is not particularly important to me. It's a small implementation detail and if you feel strongly about it I don't mind just shrugging and seeing what happens.